### PR TITLE
fix: Improve LocalDate generation

### LINKED ISSUE #3877: Arb.localDate(LocalDate.MIN, LocalDate.MAX) fails
Hi,

`Arb.localDate(LocalDate.MIN, LocalDate.MAX)` tries to build a list of all February 28s/29s, which would be approximately 1 billion dates. That's not only horribly slow, but easily leads to OutOfMemoryError.

I'm suggesting a lazy approach for calculating the edge cases and will raise a PR for it.

Best,
Oliver

### PR DESCRIPTION
**Pull request by LeoColman**
Completes #3878
Fixes #3877

### CODE REVIEW

File: kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt [hunk 1]
```diff
@@ -42,31 +43,47 @@ fun Arb.Companion.period(maxYear: Int = 10): Arb<Period> = arbitrary(listOf(Peri
...
-   minDate == maxDate -> Arb.constant(minDate)
-   else -> {
-      val leapYears = (minDate.year..maxDate.year).filter { isLeap(it.toLong()) }
+): Arb<LocalDate> {
+   require(minDate <= maxDate) { "minDate must be before or equal to maxDate" }
+   if (minDate == maxDate) return Arb.constant(minDate)
+
+   val edgeCases = mutableListOf(minDate, maxDate)
+
+   val leapYear = (minDate.year..maxDate.year).firstOrNull { isLeap(it.toLong()) }
```
Comment (AlexCue987) [inline kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Let's modify it to 
`val leapYear = (minDate.year..maxDate.year).firstOrNull { isLeap(it.toLong()) && LocalDate.of(it, 2, 29) >= minDate }`
so that we don't skip this edge case if, for instance, `minDate=LocalDate.of(2024, 3, 1)`

Do we need a unit test for this scenario?

The same goes for two other edge cases - we need to make sure edge cases are inside the range, right in the filter

File: kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt [hunk 2]
```diff
@@ -42,31 +43,47 @@ fun Arb.Companion.period(maxYear: Int = 10): Arb<Period> = arbitrary(listOf(Peri
...
+   if (leapYear != null) { edgeCases += LocalDate.of(leapYear, 2, 29) }
+
+   val centuryYear = (minDate.year..maxDate.year).firstOrNull { it % 100 == 0 }
```
Comment (AlexCue987) [inline kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Do we need unit tests to verify that these edge cases are included:
* centuryYear
* millenniumYear
* leapYear

**Review by AlexCue987 — CHANGES_REQUESTED** [review]

File: kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt (hunk 1 above)
Comment (LeoColman) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: This specific scenario is covered out by having the `removeAll` clause.

>  so that we don't skip this edge case if, for instance, minDate=LocalDate.of(2024, 3, 1)

I don't get this example, as 2024-02-29 is indeed less than 2024-03-01. Could you clarify?

File: kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt (hunk 2 above)
Comment (LeoColman) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Good idea.

We have one for leap year, but none for century and millennium. I'll include them

Comment (LeoColman) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Done. I've also removed the millennium edge cases altogether.

They don't add much value considering we already created the century edge case which will most of the times overlap with the millennium unless under very specific scenarios the first millenium is different from the first century in the date range.

File: kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt (hunk 1 above)
Comment (AlexCue987) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Suppose the range is as follows:
```kotlin
 val minDate: LocalDate = LocalDate.of(2024, 3, 1),
  val maxDate: LocalDate = LocalDate.of(2030, 12, 31)
```
the way I've grokked the PR, this means that with the code as is `leapYear == 2024`, and later `2024-02-29` is filtered out by `edgeCases.removeAll { it !in minDate..maxDate }`.

This is why I'm suggesting  a change that will compute `leapYear == 2028` and include `2028-02-29` as the edge case. Does it make sense to you?

Comment (AlexCue987) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Likewise, if the range is "2000-01-02".."2100-01-02", the centureYear edge case of "2100-01-01" is in the range., but "2000-01-01" is not. Should we provide "2100-01-01"? Please correct me if I am wrong.

Comment (LeoColman) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Ah, I see.
That's crystal clear now! I'll adjust the code

Comment (LeoColman) [reply kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/dates.kt]: Done

**Review by AlexCue987 — APPROVED** [review]
